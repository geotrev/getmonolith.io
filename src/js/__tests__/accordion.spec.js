import Undernet from "../"

const dom = `
  <div data-accordion="accordion-id" class="accordion">
    <div class="collapsible" data-collapsible="collapsible-id-1">
      <h5>
        <button id="collapsible-trigger-id" class="collapsible-trigger" data-parent="accordion-id" data-target="collapsible-id-1">
          Collapsible Trigger
        </button>
      </h5>
      <div class="collapsible-content" id="collapsible-id-1">
        <p class="has-p">
          Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
          ipsum ut voluptate. <a href="#">E pluribus unum.</a>
        </p>
      </div>
    </div>
    <div class="collapsible" data-collapsible="collapsible-id-2" data-visible="false">
      <h5>
        <button id="collapsible-trigger-id" class="collapsible-trigger" data-parent="accordion-id" data-target="collapsible-id-2">
          Collapsible Trigger
        </button>
      </h5>
      <div class="collapsible-content" id="collapsible-id-2">
        <p class="has-p">
          Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
          ipsum ut voluptate. <a href="#">E pluribus unum.</a>
        </p>
      </div>
    </div>
  </div>
`

describe("Accordion", () => {
  let wrapper

  beforeEach(() => {
    wrapper = () => document.querySelector("[data-accordion='accordion-id']")
  })

  afterEach(() => Undernet.Accordions.stop())

  describe("API start", () => {
    beforeEach(() => {
      document.body.innerHTML = dom
    })

    it("sets attributes on collapsibles", () => {
      Undernet.Accordions.start()
      expect(wrapper()).toMatchSnapshot()
    })

    it("can open multiple collapsibles initially", () => {
      // Given
      document
        .querySelector("[data-collapsible='collapsible-id-2']")
        .removeAttribute("data-visible")
      // When
      Undernet.Accordions.start()
      // Then
      expect(wrapper()).toMatchSnapshot()
    })
  })

  describe("API stop -> handleClick", () => {
    it("does not toggle collapsible", () => {
      // Given
      document.body.innerHTML = dom
      const trigger = document.querySelector("[data-target='collapsible-id-2']")
      // When
      Undernet.Accordions.start()
      Undernet.Accordions.stop()
      trigger.click()
      // Then
      expect(wrapper()).toMatchSnapshot()
    })
  })

  describe("#handleClick", () => {
    it("opens closed collapsible", () => {
      // Given
      document.body.innerHTML = dom
      const trigger = document.querySelector("[data-target='collapsible-id-2']")
      // When
      Undernet.Accordions.start()
      trigger.click()
      // Then
      expect(wrapper()).toMatchSnapshot()
    })

    it("closes open collapsible", () => {
      // Given
      document.body.innerHTML = dom
      const trigger = document.querySelector("[data-target='collapsible-id-1']")
      // When
      Undernet.Accordions.start()
      trigger.click()
      // Then
      expect(wrapper()).toMatchSnapshot()
    })

    describe("Multiple visible by default", () => {
      beforeEach(() => {
        document.body.innerHTML = dom
        wrapper()
          .querySelector("[data-collapsible='collapsible-id-2']")
          .removeAttribute("data-visible")
      })

      describe("open collapsible click", () => {
        it("keeps other collapsibles open", () => {
          // Given
          const trigger = document.querySelector("[data-target='collapsible-id-1']")
          Undernet.Accordions.start()
          // When
          trigger.click()
          // Then
          expect(wrapper()).toMatchSnapshot()
        })
      })

      describe("closed collapsible click", () => {
        it("closes other collapsibles", () => {
          // Given
          const trigger = document.querySelector("[data-target='collapsible-id-1']")
          Undernet.Accordions.start()
          // When
          // close collapsible
          trigger.click()
          // re-open collapsible
          trigger.click()
          // Then
          expect(wrapper()).toMatchSnapshot()
        })
      })
    })
  })

  describe("#setToggleMultiple", () => {
    beforeEach(() => {
      document.body.innerHTML = dom
      wrapper().setAttribute("data-toggle-multiple", null)
    })

    it("retains visible state of unselected collapsibles on click", () => {
      // Given
      const trigger = document.querySelector("[data-target='collapsible-id-2']")
      Undernet.Accordions.start()
      // When
      trigger.click()
      // Then
      expect(wrapper()).toMatchSnapshot()
    })
  })
})

const errorDom = (accordionId = "", parentId = "") => `
  <div data-accordion="${accordionId}" class="accordion">
    <div class="collapsible" data-collapsible="collapsible-id-1">
      <h5>
        <button id="collapsible-trigger-id" class="collapsible-trigger" data-parent="${parentId}" data-target="collapsible-id-1">
          Collapsible Trigger
        </button>
      </h5>
      <div class="collapsible-content" id="collapsible-id-1">
        <p class="has-p">
          Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
          ipsum ut voluptate. <a href="#">E pluribus unum.</a>
        </p>
      </div>
    </div>
    <div class="collapsible" data-collapsible="collapsible-id-2">
      <h5>
        <button id="collapsible-trigger-id" class="collapsible-trigger" data-parent="accordion-id" data-target="collapsible-id-2">
          Collapsible Trigger
        </button>
      </h5>
      <div class="collapsible-content" id="collapsible-id-2">
        <p class="has-p">
          Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
          ipsum ut voluptate. <a href="#">E pluribus unum.</a>
        </p>
      </div>
    </div>
  </div>
`

describe("Accordion Error Handling", () => {
  beforeEach(() => {
    console.warn = jest.fn()
  })

  afterEach(() => {
    Undernet.Accordions.stop()
  })

  it("logs warning if accordion id can't be found", () => {
    // Given
    const noIdWarning =
      "Could not initialize accordion; you must include a value for the 'data-accordion' attribute."
    document.body.innerHTML = errorDom("accordion-id", "")
    const trigger = document.querySelector("[data-target='collapsible-id-1']")
    // When
    Undernet.Accordions.start()
    trigger.click()
    // Then
    expect(console.warn).toBeCalledWith(noIdWarning)
  })

  it("logs warning if accordion element can't be found", () => {
    // Given
    const noAccordionWarning = "Could not find element matching [data-accordion='accordion-id']"
    document.body.innerHTML = errorDom("", "accordion-id")
    const trigger = document.querySelector("[data-target='collapsible-id-1']")
    // When
    Undernet.Accordions.start()
    trigger.click()
    // Then
    expect(console.warn).toBeCalledWith(noAccordionWarning)
  })
})
