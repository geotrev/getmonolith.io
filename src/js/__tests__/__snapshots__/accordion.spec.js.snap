// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordions #render -> Accordion Button Click (no toggle-multiple) Button content is expanded opens clicked accordion and closes all others 1`] = `
<body>
  
  
  <div
    class="accordion"
    data-accordion="accordion-1"
  >
    
    
    <div
      class="accordion-row"
      data-accordion-row="content-1"
      data-visible="false"
    >
      
      
      <h5>
        
        
        <button
          aria-controls="content-1"
          aria-expanded="false"
          class="accordion-button"
          data-parent="accordion-1"
          data-target="content-1"
          id="button-1"
        >
          
          Accordion Button 1
        
        </button>
        
      
      </h5>
      
      
      <div
        aria-hidden="true"
        aria-labelledby="button-1"
        class="accordion-content"
        id="content-1"
        style=""
      >
        
        
        <p
          class="has-margin"
        >
          
          Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
          ipsum ut voluptate. 
          <a
            href="#"
            tabindex="-1"
          >
            E pluribus unum.
          </a>
          
        
        </p>
        
      
      </div>
      
    
    </div>
    
    
    <div
      class="accordion-row"
      data-accordion-row="content-2"
      data-visible="true"
    >
      
      
      <h5>
        
        
        <button
          aria-controls="content-2"
          aria-expanded="true"
          class="accordion-button"
          data-parent="accordion-1"
          data-target="content-2"
          id="button-2"
        >
          
          Accordion Button 2
        
        </button>
        
      
      </h5>
      
      
      <div
        aria-hidden="false"
        aria-labelledby="button-2"
        class="accordion-content"
        id="content-2"
        style="max-height: 0px;"
      >
        
        
        <p
          class="has-margin"
        >
          
          Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
          ipsum ut voluptate. 
          <a
            href="#"
            tabindex="0"
          >
            E pluribus unum.
          </a>
          
        
        </p>
        
      
      </div>
      
    
    </div>
    
  
  </div>
  

</body>
`;

exports[`Accordions API start sets attributes 1`] = `
<body>
  
  
  <div
    class="accordion"
    data-accordion="accordion-1"
  >
    
    
    <div
      class="accordion-row"
      data-accordion-row="content-1"
      data-visible="true"
    >
      
      
      <h5>
        
        
        <button
          aria-controls="content-1"
          aria-expanded="true"
          class="accordion-button"
          data-parent="accordion-1"
          data-target="content-1"
          id="button-1"
        >
          
          Accordion Button 1
        
        </button>
        
      
      </h5>
      
      
      <div
        aria-hidden="false"
        aria-labelledby="button-1"
        class="accordion-content"
        id="content-1"
        style="max-height: 0px;"
      >
        
        
        <p
          class="has-margin"
        >
          
          Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
          ipsum ut voluptate. 
          <a
            href="#"
            tabindex="0"
          >
            E pluribus unum.
          </a>
          
        
        </p>
        
      
      </div>
      
    
    </div>
    
    
    <div
      class="accordion-row"
      data-accordion-row="content-2"
      data-visible="false"
    >
      
      
      <h5>
        
        
        <button
          aria-controls="content-2"
          aria-expanded="false"
          class="accordion-button"
          data-parent="accordion-1"
          data-target="content-2"
          id="button-2"
        >
          
          Accordion Button 2
        
        </button>
        
      
      </h5>
      
      
      <div
        aria-hidden="true"
        aria-labelledby="button-2"
        class="accordion-content"
        id="content-2"
      >
        
        
        <p
          class="has-margin"
        >
          
          Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
          ipsum ut voluptate. 
          <a
            href="#"
            tabindex="-1"
          >
            E pluribus unum.
          </a>
          
        
        </p>
        
      
      </div>
      
    
    </div>
    
  
  </div>
  

</body>
`;

exports[`Accordions API stop -> Accordion Button Click does not set attributes 1`] = `
<div
  aria-hidden="true"
  aria-labelledby="button-2"
  class="accordion-content"
  id="content-2"
>
  
        
  <p
    class="has-margin"
  >
    
          Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
          ipsum ut voluptate. 
    <a
      href="#"
      tabindex="-1"
    >
      E pluribus unum.
    </a>
    
        
  </p>
  
      
</div>
`;

exports[`Accordions Toggle Multiple -> Accordion Button Click opens clicked accordion and doesn't close the others 1`] = `
<body>
  
  
  <div
    class="accordion"
    data-accordion="accordion-1"
    data-toggle-multiple="null"
  >
    
    
    <div
      class="accordion-row"
      data-accordion-row="content-1"
      data-visible="true"
    >
      
      
      <h5>
        
        
        <button
          aria-controls="content-1"
          aria-expanded="true"
          class="accordion-button"
          data-parent="accordion-1"
          data-target="content-1"
          id="button-1"
        >
          
          Accordion Button 1
        
        </button>
        
      
      </h5>
      
      
      <div
        aria-hidden="false"
        aria-labelledby="button-1"
        class="accordion-content"
        id="content-1"
        style="max-height: 0px;"
      >
        
        
        <p
          class="has-margin"
        >
          
          Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
          ipsum ut voluptate. 
          <a
            href="#"
            tabindex="0"
          >
            E pluribus unum.
          </a>
          
        
        </p>
        
      
      </div>
      
    
    </div>
    
    
    <div
      class="accordion-row"
      data-accordion-row="content-2"
      data-visible="true"
    >
      
      
      <h5>
        
        
        <button
          aria-controls="content-2"
          aria-expanded="true"
          class="accordion-button"
          data-parent="accordion-1"
          data-target="content-2"
          id="button-2"
        >
          
          Accordion Button 2
        
        </button>
        
      
      </h5>
      
      
      <div
        aria-hidden="false"
        aria-labelledby="button-2"
        class="accordion-content"
        id="content-2"
        style="max-height: 0px;"
      >
        
        
        <p
          class="has-margin"
        >
          
          Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
          ipsum ut voluptate. 
          <a
            href="#"
            tabindex="0"
          >
            E pluribus unum.
          </a>
          
        
        </p>
        
      
      </div>
      
    
    </div>
    
  
  </div>
  

</body>
`;
